# Data analyst test task Besample
 
You have a [list of conference participants](https://docs.google.com/spreadsheets/d/1cYFxEhuJBt4fJNrrLgbF4LblzQ9Ru_9nj6Yyxa7FmCQ/edit?usp=sharing). We want to find all the participants who work in academia. Filter the conference participants list so that:

- you have there only people who has left emails that belong to university domains;
- the table includes the name of the university that the email belongs to

Additional but not necessary:

- table includes the rank of the university in the World University Rankings 1.2 if it has any

Useful links:

- [https://github.com/Hipo/university-domains-list](https://github.com/Hipo/university-domains-list)
- [World University Rankings 1.2](https://docs.google.com/spreadsheets/d/19OKuHOPf813i0_FqnsVlgRjDj4-mrFSi/edit?usp=sharing&ouid=114261480763518747573&rtpof=true&sd=true)

In [1]:
import requests
import json
import pandas as pd
import numpy as np

## Основная таблица `list of conference participants`

In [2]:
df = pd.read_csv('/home/asketon/Documents/Besample/APA List Main - List.csv')

In [3]:
df.head()

torosyan.adam@gmail.com  \
0                   NATIVIDAD MEDICAL CENTER   
1  COMMUNITY SERVICE BOARD OF MIDDLE GEORGIA   
2         COOK CHILDREN'S HEALTH CARE SYSTEM   
3                             ALTRU HOSPITAL   
4                  NUFFIELD HOSPITAL TAUNTON   

                           Web address  Contact Name First Name Middle Name  \
0             http://www.natividad.com     A Pascual          A          --   
1                 http://www.csbmg.com  A J Giannini          A           J   
2         http://www.cookchildrens.org      A Winter          A          --   
3                 http://www.altru.org      A Cooley          A          --   
4  http://www.nuffieldhospitals.org.uk    A M Easton        A M          --   

  Last Name         Title                   Address            Address1  \
0   Pascual  Psychologist    1441 Constitution Blvd                  --   
1  Giannini  Psychologist  2121a Bellevue Rd Bldg 7                  --   
2    Winter  Psychologist        801 Seventh Avenue                  --   
3    Cooley  Psychologist  1200 South Columbia Road                  --   
4    Easton  Psychologist   Nuffield House, 1-4 The  Crescent, Surbiton   

          City    State    ZIP Code         Country   Phone Number Employees  \
0      Salinas       CA       93915             USA   831-755-4111        --   
1       Dublin       GA  31021-9040             USA   478-272-1190       369   
2   Fort Worth       TX       76104             USA   682-885-4555      2000   
3  Grand Forks       ND       58208             USA   701-780-5200        --   
4       London  ENGLAND     KT6 4BN  United Kingdom  44-1926427971       225   

   Revenue    Industry                              Email  
0       --  Healthcare    stephanie_pascual@natividad.com  
1    $0.95  Healthcare                agiannini@csbmg.com  
2  $121.20  Healthcare               aw@cookchildrens.org  
3       --  Healthcare             marvincooley@altru.org  
4    $5.07  Healthcare  a.easton@nuffieldhospitals.org.uk

Вероятно в столбце `torosyan.adam@gmail.com` наименования учреждений, заполненные участниками конференции

In [4]:
df.rename(columns = {'torosyan.adam@gmail.com':'free_description_company'}, inplace=True)

In [5]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 15409 entries, 0 to 15408
Data columns (total 18 columns):
 #   Column                    Non-Null Count  Dtype 
---  ------                    --------------  ----- 
 0   free_description_company  15409 non-null  object
 1   Web address               15409 non-null  object
 2   Contact Name              15409 non-null  object
 3   First Name                15409 non-null  object
 4   Middle Name               15409 non-null  object
 5   Last Name                 15409 non-null  object
 6   Title                     15409 non-null  object
 7   Address                   15409 non-null  object
 8   Address1                  15409 non-null  object
 9   City                      15409 non-null  object
 10  State                     15409 non-null  object
 11  ZIP Code                  15409 non-null  object
 12  Country                   15409 non-null  object
 13  Phone Number              15409 non-null  object
 14  Employees             

In [6]:
# Создание столбцов с доменами из email
df['email_domain'] = df['Email'].apply(lambda x: x.split('@')[-1])
df['top_level_domain'] = df['Email'].apply(lambda x: x.split('.')[-1])

In [7]:
df.describe()

free_description_company Web address Contact Name First Name  \
count                        15409       15409        15409      15409   
unique                        5203        3659        15083       4554   
top     KAISER PERMANENTE HOSPITAL          --    Jon Allen       John   
freq                           306         931            3        313   

       Middle Name Last Name         Title                 Address Address1  \
count        15409     15409         15409                   15409    15409   
unique         779     10207             1                    6333      528   
top             --     Smith  Psychologist  1 Gustave L Levy Place       --   
freq         11523        65         15409                     262    14243   

            City  State    ZIP Code Country  Phone Number Employees Revenue  \
count      15409  15409       15409   15409         15409     15409   15409   
unique      1959     93        5204      11          6122       812    1495   
top     New York     TX  10029-6574     USA  212-241-6500        --      --   
freq         658   1699         342   14792           359      5823    5984   

          Industry                            Email email_domain  \
count        15409                            15409        15409   
unique           1                            15409         3560   
top     Healthcare  stephanie_pascual@natividad.com      aol.com   
freq         15409                                1          468   

       top_level_domain  
count             15409  
unique               17  
top                 org  
freq               6909

In [8]:
df['top_level_domain'].value_counts()

top_level_domain
org     6909
com     5036
edu     2291
ca       394
net      388
us       236
uk        78
au        46
info      13
mx         5
cc         3
jp         2
be         2
md         2
de         2
ie         1
in         1
Name: count, dtype: int64

Описание окончаний доменных имен верхнего уровня (Top-Level Domains, TLDs)
- .org (Organization): Обычно используется некоммерческими организациями, общественными объединениями и фондами. Однако с течением времени стал использоваться и коммерческими организациями.

- .com (Commercial): Самый распространенный и широко используемый TLD. Предназначен для коммерческих организаций, но не ограничивается ими.

- .edu (Education): Используется образовательными учреждениями, такими как университеты, колледжи.

- .ca (Canada): Представляет Канаду и является страновым доменом верхнего уровня для этой страны.

- .net (Network): Изначально предназначен для сетевых организаций, но с течением времени использование расширилось.

- .us (United States): Страновой домен верхнего уровня для Соединенных Штатов Америки.

- .uk (United Kingdom): Страновой домен верхнего уровня для Великобритании.

- .au (Australia): Страновой домен верхнего уровня для Австралии.

- .info (Information): Используется для информационных ресурсов.

- .mx (Mexico): Страновой домен верхнего уровня для Мексики.

- .cc (Cocos Islands): Доступен для регистрации без ограничений, часто используется как альтернатива ".com".

.edu — общий домен верхнего уровня для образовательных учреждений. «Домен предназначен для аккредитованных высших учебных заведений США», и это намерение строго соблюдается. 

In [9]:
df[df['top_level_domain'] == 'edu'].describe()

free_description_company           Web address       Contact Name  \
count                      2291                  2291               2291   
unique                      567                   446               2272   
top            BROWN UNIVERSITY  http://www.brown.edu  Alexandre Todorov   
freq                        176                   176                  2   

       First Name Middle Name Last Name         Title                 Address  \
count        2291        2291      2291          2291                    2291   
unique       1082         224      1983             1                     730   
top       Michael          --    Miller  Psychologist  1 Gustave L Levy Place   
freq           38        1558         6          2291                     237   

       Address1      City State    ZIP Code Country  Phone Number Employees  \
count      2291      2291  2291        2291    2291          2291      2291   
unique       54       313    49         615       2           715       182   
top          --  New York    NY  10029-6574     USA  212-241-6500        --   
freq       2205       318   443         240    2286           256       626   

       Revenue    Industry                     Email email_domain  \
count     2291        2291                      2291         2291   
unique     204           1                      2291          414   
top         --  Healthcare  aalok.turakhia@emory.edu     mssm.edu   
freq       639        2291                         1          261   

       top_level_domain  
count              2291  
unique                1  
top                 edu  
freq               2291

проверка на дубликаты:

In [10]:
df.duplicated().sum()

0

### Выводы:

В таблице 15408 строк. Судя по доменам, у нас до 3560 компаний зарегистрировано на конференцию. 2291 строк относится к 414 доменам аккредитованным высшим учебным заведениям США с окончаниием домена `.edu` (15% от всех участников). Мы можем использовать это для нашей задачи.

## Таблица `university-domains-list` с Hipo

In [11]:
requests.get("http://universities.hipolabs.com/search?name=KAISER").text

'[{"alpha_two_code": "DE", "domains": ["uni-kl.de"], "web_pages": ["http://www.uni-kl.de/"], "name": "Universit\\u00e4t Kaiserslautern", "state-province": null, "country": "Germany"}, {"alpha_two_code": "DE", "domains": ["fh-kl.de"], "web_pages": ["http://www.fh-kl.de/"], "name": "Fachhochschule Kaiserslautern", "state-province": null, "country": "Germany"}]'

In [12]:
api_response = requests.get("http://universities.hipolabs.com/search?")
print(api_response.status_code)

200


In [13]:
# Проверка успешности запроса
if api_response.status_code == 200:
    # Извлечение текстового содержимого ответа
    response_text = api_response.text

    # Преобразование текстового содержимого в формат JSON
    data = json.loads(response_text)

    # Создание DataFrame
    df_univer = pd.DataFrame(data)
else:
    print(f"Ошибка при запросе API. Статус код: {api_response.status_code}")

In [14]:
df_univer.head(2)

alpha_two_code                           domains             web_pages  \
0             UA  [student.karazin.ua, karazin.ua]  [https://karazin.ua]   
1             CL                          [usm.cl]      [https://usm.cl]   

                                       name state-province  country  
0               Kharkiv National University           None  Ukraine  
1  Universidad Técnica Federico Santa María           None    Chile

In [15]:
df_univer.shape

(9947, 6)

In [16]:
df_univer.describe()

alpha_two_code      domains                 web_pages  \
count            9947         9947                      9947   
unique            202         9943                      9942   
top                US  [howest.be]  [http://www.agu.edu.tr/]   
freq             2270            2                         2   

                        name state-province        country  
count                   9947           1127           9947  
unique                  9863            256            205  
top     Arab Open University     Tamil Nadu  United States  
freq                       6             56           2269

In [17]:
df_univer['alpha_two_code'].unique()

array(['UA', 'CL', 'FR', 'CN', 'GB', 'US', 'IN', 'CA', 'AF', 'AL', 'DZ',
       'AD', 'AO', 'AG', 'AR', 'AM', 'AU', 'AT', 'AZ', 'BS', 'BH', 'BD',
       'BB', 'BY', 'BE', 'BZ', 'BJ', 'BM', 'BT', 'BO', 'BA', 'BW', 'BR',
       'BN', 'BG', 'BF', 'MM', 'BI', 'KH', 'CM', 'CV', 'KY', 'CF', 'TD',
       'CO', 'CD', 'CG', 'CR', 'CI', 'IR', 'PL', 'HR', 'CU', 'CY', 'CZ',
       'DK', 'DJ', 'DM', 'DO', 'EC', 'EG', 'SV', 'GQ', 'ER', 'EE', 'ET',
       'FO', 'FJ', 'FI', 'GF', 'PF', 'GA', 'GM', 'GE', 'DE', 'GH', 'GR',
       'GL', 'GD', 'GP', 'GU', 'GT', 'GN', 'GY', 'HT', 'VA', 'HN', 'HK',
       'HU', 'IS', 'PK', 'ID', 'IQ', 'IE', 'IL', 'IT', 'JM', 'JP', 'JO',
       'KZ', 'KE', 'KP', 'KR', 'XK', 'KW', 'KG', 'LA', 'LV', 'LB', 'LS',
       'LR', 'LY', 'LI', 'LT', 'LU', 'MO', 'MK', 'MG', 'MW', 'MY', 'MV',
       'ML', 'MT', 'MR', 'MU', 'MX', 'MD', 'MC', 'MN', 'ME', 'MS', 'MA',
       'MZ', 'NA', 'NP', 'NL', 'NC', 'NZ', 'NI', 'NE', 'NG', 'NU', 'NO',
       'OM', 'PS', 'PA', 'PG', 'PY', 'PE', 'PH', 'P

In [18]:
# Разделение доменов в столбце 'domains'
df_univer = df_univer.explode('domains')

In [19]:
df_univer.head(2)

alpha_two_code             domains             web_pages  \
0             UA  student.karazin.ua  [https://karazin.ua]   
0             UA          karazin.ua  [https://karazin.ua]   

                          name state-province  country  
0  Kharkiv National University           None  Ukraine  
0  Kharkiv National University           None  Ukraine

In [20]:
df_univer.shape

(10193, 6)

In [21]:
df_univer[df_univer['domains'].str.count('\.') >= 2]

alpha_two_code             domains  \
0                UA  student.karazin.ua   
3                CN   mail2.sysu.edu.cn   
3                CN    mail.sysu.edu.cn   
4                GB          rhul.ac.uk   
5                CN          hnu.edu.cn   
...             ...                 ...   
9937             PT    www.esmad.ipp.pt   
9941             TR          thk.edu.tr   
9942             PL    if-pan.krakow.pl   
9943             MX        lamar.org.mx   
9944             MA          ehtp.ac.ma   

                                              web_pages  \
0                                  [https://karazin.ua]   
3                                 [https://sysu.edu.cn]   
3                                 [https://sysu.edu.cn]   
4                                  [https://rhul.ac.uk]   
5     [https://www.hnu.edu.cn, http://www-en.hnu.edu...   
...                                                 ...   
9937                        [https://www.esmad.ipp.pt/]   
9941                              [https://thk.edu.tr/]   
9942                        [https://if-pan.krakow.pl/]   
9943                                [https://lamar.mx/]   
9944                           [http://www.ehtp.ac.ma/]   

                                                   name state-province  \
0                           Kharkiv National University           None   
3                                Sun Yat-Sen University           None   
3                                Sun Yat-Sen University           None   
4                   Royal Holloway University of London           None   
5                                      Hunan University           None   
...                                                 ...            ...   
9937            Escola Superior de Media Artes e Design           None   
9941                      Türk Hava Kurumu Üniversitesi           None   
9942  Maj Institute of Pharmacology Polish Academy o...           None   
9943                      Universidad Guadalajara Lamar           None   
9944                 Ecole Hassania des Travaux Publics           None   

             country  
0            Ukraine  
3              China  
3              China  
4     United Kingdom  
5              China  
...              ...  
9937        Portugal  
9941          Turkey  
9942          Poland  
9943          Mexico  
9944         Morocco  

[5230 rows x 6 columns]

In [22]:
df_univer['top_level_domain'] = df_univer['domains'].apply(lambda x: x.split('.')[-1])

In [23]:
df_univer.describe()

alpha_two_code            domains  \
count           10193              10193   
unique            202              10187   
top                US  student.sdccd.edu   
freq             2308                  3   

                                         web_pages                  name  \
count                                        10193                 10193   
unique                                        9942                  9863   
top     [https://sciences.sorbonne-universite.fr/]  Arab Open University   
freq                                             5                     6   

       state-province        country top_level_domain  
count            1176          10193            10193  
unique            256            205              194  
top        Tamil Nadu  United States              edu  
freq               58           2307             2520

9947 университета в списке http://universities.hipolabs.com, 25% в домене `.edu`

## Отбор строк с доменами академий

посмотрим на кол-во строк при полном совпадении доменов

In [24]:
# Объединение таблиц по условию вхождения доменной части
result_df = pd.merge(df, df_univer, left_on='email_domain', right_on='domains', how='inner')

In [25]:
result_df.shape

(1437, 27)

частичное совпадение

NOTE: Some universities use a format like [user]@[department].[domain], but this list only contains the [domain] portion. For example, an email address might be [student]@cs.usc.edu, and this list will contain 'usc.edu', the domain for the University of Southern California. Take this into consideration if using this list for email address validation.

In [26]:
def first_match(series, email):
    for value in series:
        if value in email:
            return value
    return None 

In [27]:
%%time
df['temp_join_column'] = df['email_domain'].apply(lambda x: first_match(df_univer['domains'], x))

CPU times: user 18.2 s, sys: 6.73 ms, total: 18.2 s
Wall time: 18.2 s


In [28]:
df['temp_join_column'].count()

2035

In [29]:
merged_df = pd.merge(df, df_univer, left_on='temp_join_column', right_on='domains', how='left')

посмотрим на кол-во присоединенных строк

In [30]:
participants_univer_short = merged_df[(merged_df['name'].notna())]
participants_univer_short.shape[0]

2035

часть универов утеряна

кол-во строк с доменом `.edu` найденных в списке `university-domains-list`:

In [31]:
merged_df[(merged_df['top_level_domain_x'] == 'edu') & (merged_df['name'].notna())].shape[0]

1890

In [32]:
merged_df[(merged_df['top_level_domain_x'] == 'edu')].shape[0]

2291

Всего строк с доменом .edu 2291, то есть не нашлось в списке university-domains-list 401 строка (17%)

отберем строки с универами с условием попадания или по `.edu` или по списку:

In [33]:
participants_univer = merged_df[(merged_df['top_level_domain_x'] == 'edu') | (merged_df['name'].notna())]

In [34]:
participants_univer.shape[0]

2436

мы можем дополнительно изучить сайты строк не попавших в список university-domains-list, чтобы определиться с выборкой. Остановимся на 2436 строках.

In [35]:
participants_univer[(participants_univer['top_level_domain_x'] == 'edu') & (participants_univer['name'].isna())].sample(5)

free_description_company  \
3770   CANCER CARE CENTERS OF SOUTHWEST TEXAS   
6926                                JEFFERSON   
947   GEISINGER WYOMING VALLEY MEDICAL CENTER   
9531                                JEFFERSON   
5843                             PACKER IRA K   

                              Web address   Contact Name First Name  \
3770  http://www.cancercaresouthtexas.com   Dina Chavira       Dina   
6926        http://www.wearejefferson.com  Jingduan Yang   Jingduan   
947              http://www.geisinger.org     April Rine      April   
9531        http://www.wearejefferson.com  Margaret Fuhs   Margaret   
5843              http://www.umassmed.edu   Ira K Packer        Ira   

     Middle Name Last Name         Title  \
3770          --   Chavira  Psychologist   
6926          --      Yang  Psychologist   
947           --      Rine  Psychologist   
9531          --      Fuhs  Psychologist   
5843           K    Packer  Psychologist   

                                         Address      Address1          City  \
3770                      40 Ne Loop 410 Ste 500            --   San Antonio   
6926                       111 Sourh 11th Street    Suite 4260  Philadelphia   
947                         100 N Academy Avenue            --      Danville   
9531                               5060 State Rd  Second Floor   Drexel Hill   
5843  University Of Massachusetts Medical School            --     Worcester   

      ...   email_domain top_level_domain_x temp_join_column alpha_two_code  \
3770  ...    uthscsa.edu                edu             None            NaN   
6926  ...  jefferson.edu                edu             None            NaN   
947   ...  geisinger.edu                edu             None            NaN   
9531  ...  jefferson.edu                edu             None            NaN   
5843  ...   umassmed.edu                edu             None            NaN   

     domains web_pages name state-province country top_level_domain_y  
3770     NaN       NaN  NaN            NaN     NaN                NaN  
6926     NaN       NaN  NaN            NaN     NaN                NaN  
947      NaN       NaN  NaN            NaN     NaN                NaN  
9531     NaN       NaN  NaN            NaN     NaN                NaN  
5843     NaN       NaN  NaN            NaN     NaN                NaN  

[5 rows x 28 columns]

In [36]:
participants_univer.columns

Index(['free_description_company', 'Web address', 'Contact Name', 'First Name',
       'Middle Name', 'Last Name', 'Title', 'Address', 'Address1', 'City',
       'State', 'ZIP Code', 'Country', 'Phone Number', 'Employees', 'Revenue',
       'Industry', 'Email', 'email_domain', 'top_level_domain_x',
       'temp_join_column', 'alpha_two_code', 'domains', 'web_pages', 'name',
       'state-province', 'country', 'top_level_domain_y'],
      dtype='object')

In [37]:
participants_univer_short.email_domain.nunique()

371

In [38]:
participants_univer_short.name.nunique()

272

In [39]:
participants_univer_short.domains.nunique()

274

Некоторым университетам принадлежит несколько доменных имен

In [40]:
participants_univer.email_domain.nunique()

440

После добавления фильтра по университетам, кол-во строк уменьшилось с 15409 до 2436, доменных имен с 3560 до 440.

## Добавление информации по рангам

### Изучим таблицу `World University Rankings`

In [41]:
df_rank = pd.read_csv('/home/asketon/Documents/Besample/2024 QS World University Rankings 1.2.csv', skiprows=1)

In [42]:
df_rank = df_rank.drop([0, 1])

In [43]:
df_rank.head(3)

2024 2023                              Institution Name Location  \
2    1    1  Massachusetts Institute of Technology (MIT)        US   
3    2    2                       University of Cambridge       UK   
4    3    4                          University of Oxford       UK   

       Unnamed: 4 Classification Unnamed: 6 Unnamed: 7 Unnamed: 8  \
2   United States              M         CO         VH          B   
3  United Kingdom              L         FC         VH          A   
4  United Kingdom              L         FC         VH          A   

  Academic Reputation  ... Unnamed: 18 International Students Unnamed: 20  \
2               100.0  ...          56                   88.2         128   
3               100.0  ...          64                   95.8          85   
4               100.0  ...         110                   98.2          60   

  International Research Network Unnamed: 22 Employment Outcomes Unnamed: 24  \
2                           94.3          58                 100           4   
3                           99.9           7                 100           6   
4                          100.0           1                 100           3   

  Sustainability Unnamed: 26 Overall  
2           95.2          51   100.0  
3           97.3         33=    99.2  
4           97.8         26=    98.9  

[3 rows x 28 columns]

In [44]:
df_rank.describe()

2024      2023                              Institution Name  \
count        1498      1415                                          1498   
unique        380       333                                          1498   
top     1201-1400  801-1000  Massachusetts Institute of Technology (MIT)    
freq          206       204                                             1   

       Location     Unnamed: 4 Classification Unnamed: 6 Unnamed: 7  \
count      1498           1498           1475       1497       1411   
unique      104            104              4          4          4   
top          US  United States              L         FC         VH   
freq        199            199            689        596        923   

       Unnamed: 8 Academic Reputation  ... Unnamed: 18 International Students  \
count        1453                1498  ...        1372                   1418   
unique          3                 477  ...         681                    524   
top             A                 5.7  ...        701+                    1.1   
freq         1156                  21  ...         685                     36   

       Unnamed: 20 International Research Network Unnamed: 22  \
count         1418                           1494        1494   
unique         685                            553         699   
top           701+                            1.0        701+   
freq           728                            281         791   

       Employment Outcomes Unnamed: 24 Sustainability Unnamed: 26 Overall  
count                 1474        1474           1398        1398    1498  
unique                 449         690            399         343     352  
top                    7.9        701+              1        701+       -  
freq                    20         782            174         698     896  

[4 rows x 28 columns]

В таблице 1498 университетов

### Присоединим информацию по рангам к нашей таблице

In [45]:
participants_univer = participants_univer.copy()
participants_univer['name'].fillna('unknown', inplace=True)

In [46]:
participants_univer['merged_name'] = participants_univer['name'].str.lower().apply(lambda x: x.split('(')[0])
df_rank['merged_name'] = df_rank['Institution Name'].str.lower().apply(lambda x: x.split('(')[0])

In [47]:
df_final = pd.merge(participants_univer, df_rank[['2024', 'merged_name']], left_on='merged_name', right_on='merged_name', how='left')

In [48]:
df_final.head()

free_description_company  \
0                                   EMORY HEALTHCARE   
1             CANCER CARE CENTERS OF SOUTHWEST TEXAS   
2                     NORTHWESTERN MEMORIAL HOSPITAL   
3  UNIVERSITY OF TEXAS HEALTH SCIENCE CENTER AT S...   
4           UNIVERSITY OF IOWA HOSPITALS AND CLINICS   

                           Web address          Contact Name First Name  \
0       http://www.emoryhealthcare.org  Aalok Bipin Turakhia      Aalok   
1  http://www.cancercaresouthtexas.com     Aaron Cobarruvias      Aaron   
2                   http://www.nmh.org      Aaron S Reichlin      Aaron   
3               http://www.uthscsa.edu       Aaron P Edwards      Aaron   
4                 http://www.uiowa.edu           Aaron Kauer      Aaron   

  Middle Name    Last Name         Title                Address Address1  \
0       Bipin     Turakhia  Psychologist     1364 Clifton Rd Ne       --   
1          --  Cobarruvias  Psychologist     7703 Floyd Curl Dr       --   
2           S     Reichlin  Psychologist     251 E Huron Street       --   
3           P      Edwards  Psychologist  7703 Floyd Curl Drive       --   
4          --        Kauer  Psychologist      200 Hawkins Drive       --   

          City  ...  temp_join_column alpha_two_code           domains  \
0      Atlanta  ...         emory.edu             US         emory.edu   
1  San Antonio  ...              None            NaN               NaN   
2      Chicago  ...  northwestern.edu             US  northwestern.edu   
3  San Antonio  ...              None            NaN               NaN   
4    Iowa City  ...         uiowa.edu             US         uiowa.edu   

                        web_pages                     name state-province  \
0         [http://www.emory.edu/]         Emory University           None   
1                             NaN                  unknown            NaN   
2  [http://www.northwestern.edu/]  Northwestern University           None   
3                             NaN                  unknown            NaN   
4         [http://www.uiowa.edu/]       University of Iowa           None   

         country top_level_domain_y              merged_name  2024  
0  United States                edu         emory university   166  
1            NaN                NaN                  unknown   NaN  
2  United States                edu  northwestern university   =47  
3            NaN                NaN                  unknown   NaN  
4  United States                edu       university of iowa  =491  

[5 rows x 30 columns]

In [49]:
df_final['2024'].sort_values().unique()

array(['1001-1200', '104', '11', '111', '112', '118', '12', '1201-1400',
       '13', '136', '16', '166', '168', '205', '23', '261', '28', '3',
       '30', '4', '42', '57', '63', '631-640', '641-650', '671-680',
       '681-690', '691-700', '721-730', '751-760', '761-770', '791-800',
       '801-850', '83', '851-900', '901-950', '951-1000', '97', '=189',
       '=237', '=340', '=375', '=441', '=469', '=47', '=491', '=498',
       '=545', '=548', '=563', '=575', '=593', '=600', '=73', '=93', nan],
      dtype=object)

In [50]:
df_final[df_final['2024'].notna()]['name'].count()

905

In [51]:
df_final[df_final['2024'].isna()]['name'].nunique()

203

In [52]:
df_final[df_final['2024'].notna()]['name'].nunique()

70

In [53]:
df_final[df_final['2024'].isna()].describe()

free_description_company          Web address   Contact Name  \
count                               1531                 1531           1531   
unique                               404                  323           1513   
top     MOUNT SINAI MEDICAL CENTER, INC.  http://www.msmc.com  Pamela Madden   
freq                                 148                  148              2   

       First Name Middle Name Last Name         Title                 Address  \
count        1531        1531      1531          1531                    1531   
unique        811         133      1360             1                     524   
top       Michael          --     Smith  Psychologist  1 Gustave L Levy Place   
freq           24        1153         5          1531                     237   

       Address1      City  ... temp_join_column alpha_two_code   domains  \
count      1531      1531  ...             1130           1130      1130   
unique       35       267  ...              202              6       202   
top          --  New York  ...         mssm.edu             US  mssm.edu   
freq       1484       292  ...              261           1109       261   

                       web_pages     name state-province        country  \
count                       1130     1531             42           1130   
unique                       202      203             12              6   
top     [http://icahn.mssm.edu/]  unknown   Pennsylvania  United States   
freq                         261      401             15           1109   

       top_level_domain_y merged_name 2024  
count                1130        1531    0  
unique                  4         203    0  
top                   edu     unknown  NaN  
freq                 1114         401  NaN  

[4 rows x 30 columns]

У нас получилось подтянуть информацию по рангам для 905 участников (70 универститетов, для 203 универститетов совпадений не нашлось)

удалим лишние столбцы

In [54]:
columns_to_drop = ['temp_join_column', 'email_domain', 'top_level_domain_x', 'domains', 'web_pages', 'state-province', 'country', 'top_level_domain_y', 'merged_name']

df_final.drop(columns=columns_to_drop, inplace=True)

In [55]:
df_final.rename(columns={'2024': 'rank'}, inplace=True)

In [56]:
df_final.head(3)

free_description_company  \
0                        EMORY HEALTHCARE   
1  CANCER CARE CENTERS OF SOUTHWEST TEXAS   
2          NORTHWESTERN MEMORIAL HOSPITAL   

                           Web address          Contact Name First Name  \
0       http://www.emoryhealthcare.org  Aalok Bipin Turakhia      Aalok   
1  http://www.cancercaresouthtexas.com     Aaron Cobarruvias      Aaron   
2                   http://www.nmh.org      Aaron S Reichlin      Aaron   

  Middle Name    Last Name         Title             Address Address1  \
0       Bipin     Turakhia  Psychologist  1364 Clifton Rd Ne       --   
1          --  Cobarruvias  Psychologist  7703 Floyd Curl Dr       --   
2           S     Reichlin  Psychologist  251 E Huron Street       --   

          City  ... ZIP Code Country  Phone Number Employees    Revenue  \
0      Atlanta  ...    30322     USA  404-712-2000        --         --   
1  San Antonio  ...    78229     USA  210-567-7000     10000     $50.00   
2      Chicago  ...    60611     USA  312-926-2000      7000  $1,290.00   

     Industry                        Email alpha_two_code  \
0  Healthcare     aalok.turakhia@emory.edu             US   
1  Healthcare     cobarruviasa@uthscsa.edu            NaN   
2  Healthcare  a-reichlin@northwestern.edu             US   

                      name rank  
0         Emory University  166  
1                  unknown  NaN  
2  Northwestern University  =47  

[3 rows x 21 columns]

## Выводы

- У нас 15408 участников конференции, из них от университетов 2436. 
- Информация по рангу их университета определена для 905 участников.